### ISSUE TITLE
Strange behavior of IncrementalSfM (Resection part) when using GPS priors (-P)

### ISSUE THREAD

**Issue by hcjghr** [openMVG/openMVG#1255]
Hi,

From the beginning I apologize for a bit longer post:) 

In recent weeks I have been noticing that the IncrementalSfM has been acting strange in the resection step when motion prior switch (-P) is set. After the first BA with motion priors resection thresholds increase significantly causing problems in subsequent steps.

To illustrate the problem I have run the same IncrementalSfM command (with and without -P) on the same feature and match files.

In both cases the process starts (nearly) the same, with same initial pair (forced by -a and -b) and then followed by a series of robust resections for new images. The robust resection threshold is varies between the images but is reasonable (around 3-4) while final RMSE after the single pose BA below 1px.

The behavior is the similar in both cases until the first complete BA (after 19 images for example with P and after 25 images without P)

**- with P**

Bundle Adjustment statistics (approximated RMSE):
 #views: 174
 #poses: 19
 #intrinsics: 1
 #tracks: 50840
 #residuals: 438129
 Initial RMSE: 0.305092
 Final RMSE: 0.188702
 Time (s): 17.3881

Usable motion priors: 1
Pose prior statistics (user units):
 - Starting median fitting error: 0.0649667
 - Final fitting error:
	 min: 0.0315291
	 mean: 0.096793
	 median: 0.0699325
	 max: 0.273009

**- without P**

Bundle Adjustment statistics (approximated RMSE):
 #views: 174
 #poses: 26
 #intrinsics: 1
 #tracks: 61845
 #residuals: 596120
 Initial RMSE: 0.218907
 Final RMSE: 0.214843
 Time (s): 6.6084

I'm assuming that the difference between would be due to the randomness in the AC-RANSAC step (does that make sense?). Also, as the motion priors are not exact the error in the alignment is in the expected range.

**The big difference occurs after the first BA with motion priors as the final thresholds of the robust resection are differ for an order of magnitude.**

**- with P**

-- Robust Resection of view: 39
  nfa=27.4825 inliers=7/8445 precisionNormalized=0.0269952 precision=229.827 (iter=0 ,sample=6880,1144,7649,7052,1075,8182,)
  nfa=26.3097 inliers=7/8445 precisionNormalized=0.00181329 precision=59.5649 (iter=2 ,sample=2351,1591,4619,8384,8086,8415,)
  nfa=25.1653 inliers=7/8445 precisionNormalized=0.000129922 precision=15.944 (iter=2 ,sample=2351,1591,4619,8384,8086,8415,)
  nfa=24.9485 inliers=7/8445 precisionNormalized=7.88136e-05 precision=12.4182 (iter=4 ,sample=8083,928,4100,6740,6759,2511,)
  nfa=-4.22006 inliers=677/8445 precisionNormalized=0.00883906 precision=131.51 (iter=5 ,sample=1198,41,3562,952,7733,5404,)
  nfa=-9.70317 inliers=291/8445 precisionNormalized=0.00308839 precision=77.7362 (iter=29 ,sample=3099,7209,6126,6925,4475,5617,)
  nfa=-10.8146 inliers=323/8445 precisionNormalized=0.00350706 precision=82.8378 (iter=51 ,sample=3226,3136,7166,6119,5184,6147,)
  nfa=-18.4524 inliers=393/8445 precisionNormalized=0.00429643 precision=91.6877 (iter=78 ,sample=7133,6179,4419,3189,3240,5741,)

-- Robust Resection 
-- Resection status: 1
-- #Points used for Resection: 8445
-- #Points validated by robust Resection: 393
-- Threshold: 91.6877

**- without P**

-- Robust Resection of view: 52
  nfa=-154.392 inliers=910/3831 precisionNormalized=0.0201675 precision=199.479 (iter=0 ,sample=3121,519,3470,3199,489,3711,)
  nfa=-2701.08 inliers=2730/3831 precisionNormalized=0.0137327 precision=164.607 (iter=0 ,sample=3121,519,3470,3199,489,3711,)
  nfa=-4895.44 inliers=2624/3831 precisionNormalized=0.00169636 precision=57.8535 (iter=2 ,sample=3027,911,3146,1173,1565,2593,)
  nfa=-12477.7 inliers=3778/3831 precisionNormalized=0.000143473 precision=16.825 (iter=2 ,sample=3027,911,3146,1173,1565,2593,)
  nfa=-12477.7 inliers=3778/3831 precisionNormalized=0.000143473 precision=16.825 (iter=2 ,sample=3027,911,3146,1173,1565,2593,)
  nfa=-13591.1 inliers=3756/3831 precisionNormalized=6.75269e-05 precision=11.5427 (iter=6 ,sample=3073,231,185,565,1079,3807,)
  nfa=-14884.4 inliers=3683/3831 precisionNormalized=2.36087e-05 precision=6.82507 (iter=34 ,sample=3048,1895,3274,1123,3646,2370,)
  nfa=-15053.2 inliers=3783/3831 precisionNormalized=3.02306e-05 precision=7.72314 (iter=197 ,sample=513,395,3070,2010,3264,1885,)
  nfa=-15202.3 inliers=3753/3831 precisionNormalized=2.47661e-05 precision=6.99036 (iter=371 ,sample=1272,3361,1947,825,566,2839,)

-- Robust Resection 
-- Resection status: 1
-- #Points used for Resection: 3831
-- #Points validated by robust Resection: 3753
-- Threshold: 6.99036

The difference in the magnitude of the result is always big, so I was wondering if somebody has an explanation what could be the reason for that. This causes weird things to happen as the thresholds are huge and points get discarded in the following process. 

**I have start noticing this shortly after v1.3 but I did not have the time to investigate and I assumed there was something wrong with my dataset that I used at the time. For this reason, I tried to run the IncrementalSfM from v1.2 (with motion priors using the same matches and features) and it does produce the same problem - thresholds of robust resections are comparable to the ones in v1.3 without motion priors:**

**- v1.2 with P**

Bundle Adjustment statistics (approximated RMSE):
 #views: 174
 #poses: 19
 #intrinsics: 1
 #tracks: 50847
 #residuals: 438123
 Initial RMSE: 0.294252
 Final RMSE: 0.188743
 Time (s): 10.6828

Usable motion priors: 1
Pose prior statistics (user units):
 - Starting median fitting error: 0.0640137
 - Final fitting error:
	 min: 0.0326317
	 mean: 0.0966948
	 median: 0.0694979
	 max: 0.273369

-- Robust Resection of view: 39
  nfa=-39309.3 inliers=8155/8445 precisionNormalized=3.97186e-06 precision=2.78864 (iter=0 ,sample=6880,1144,7649,)
  nfa=-39493.7 inliers=8183/8445 precisionNormalized=3.96468e-06 precision=2.78612 (iter=7 ,sample=1949,3030,5451,)
  nfa=-39578 inliers=8097/8445 precisionNormalized=3.2968e-06 precision=2.54063 (iter=9 ,sample=5482,861,4876,)
  nfa=-39623.6 inliers=8228/8445 precisionNormalized=4.15081e-06 precision=2.85077 (iter=27 ,sample=3926,2872,6686,)
  nfa=-39937.1 inliers=8135/8445 precisionNormalized=3.1851e-06 precision=2.49722 (iter=49 ,sample=8303,4467,8342,)
  nfa=-40197.4 inliers=8054/8445 precisionNormalized=2.52861e-06 precision=2.22503 (iter=60 ,sample=3560,4341,6475,)
  nfa=-40395.5 inliers=8116/8445 precisionNormalized=2.68379e-06 precision=2.29229 (iter=64 ,sample=573,3668,8071,)
  nfa=-40589.2 inliers=8062/8445 precisionNormalized=2.28306e-06 precision=2.11424 (iter=96 ,sample=2222,4537,3548,)

I've also tried the IncrementalSfM2 which works perfectly without P and does not find any inliers after the first BA with priors if I use the P option.

I have also tried it with different datasets so I dont think it is a dataset problem.

Has anybody noticed anything similar?

Thanks and again sorry for the long post!

(@pmoulon - I will send you the dataset and all the files through email as the data is not mine and cant be publicly shared).

**Comment by rjanvier** [comment]
@hcjghr could you please test with this pr https://github.com/openMVG/openMVG/pull/1253? Thanks.

**Comment by hcjghr** [comment]
Hi,

@rjanvier I have tried with your suggestion and it does not solve it. _The robust resections after the first complete BA when using motion priors give strange results (not the case in v1.2)._

As I have made a small mistake looking at the logs yesterday (actually both with and without P do first complete BA after 19 poses)  I will put here again the outputs of:
- last robust resection before the first complete BA
- BA
- first robust resection after the complete BA


**- most recent develop branch of openMVG with -P switch**

-- Robust Resection of view: 26
  nfa=-3396.83 inliers=2484/3328 precisionNormalized=0.00622307 precision=112.531 (iter=0 ,sample=2711,451,3014,2779,426,3224,)
  nfa=-4356.95 inliers=2573/3328 precisionNormalized=0.00313704 precision=79.8972 (iter=0 ,sample=2711,451,3014,2779,426,3224,)
  nfa=-4805.06 inliers=3169/3328 precisionNormalized=0.0077557 precision=125.627 (iter=0 ,sample=2711,451,3014,2779,426,3224,)
  nfa=-4933.26 inliers=3196/3328 precisionNormalized=0.00748884 precision=123.447 (iter=2 ,sample=1395,2309,2615,1690,2360,1513,)
  nfa=-4933.77 inliers=3222/3328 precisionNormalized=0.00792455 precision=126.987 (iter=6 ,sample=1610,1955,2073,3208,3073,2231,)
  nfa=-6135.84 inliers=3098/3328 precisionNormalized=0.00247992 precision=71.038 (iter=10 ,sample=9,2687,599,265,586,1337,)
  nfa=-6600.38 inliers=3218/3328 precisionNormalized=0.00237827 precision=69.5668 (iter=18 ,sample=464,442,1284,1469,1622,2206,)
  nfa=-9501.73 inliers=3189/3328 precisionNormalized=0.000270622 precision=23.4668 (iter=19 ,sample=475,1926,1288,2588,1980,197,)
  nfa=-13710.3 inliers=3208/3328 precisionNormalized=1.38334e-05 precision=5.30561 (iter=23 ,sample=2181,1386,842,2406,2917,1035,)
  nfa=-14224.2 inliers=3232/3328 precisionNormalized=1.05654e-05 precision=4.63676 (iter=165 ,sample=580,1687,2619,1685,221,2828,)
  nfa=-14299.2 inliers=3211/3328 precisionNormalized=9.13517e-06 precision=4.31151 (iter=278 ,sample=268,1375,1869,3072,1789,734,)
  nfa=-14967.3 inliers=3230/3328 precisionNormalized=6.11154e-06 precision=3.52652 (iter=330 ,sample=453,817,2433,556,154,3033,)

-- Robust Resection 
-- Resection status: 1
-- #Points used for Resection: 3328
-- #Points validated by robust Resection: 3230
-- Threshold: 3.52652

Bundle Adjustment statistics (approximated RMSE):
 #views: 1
 #poses: 1
 #intrinsics: 1
 #tracks: 3230
 #residuals: 6460
 Initial RMSE: 0.768685
 Final RMSE: 0.457455
 Time (s): 0.0710752

Bundle Adjustment statistics (approximated RMSE):
 #views: 174
 #poses: 19
 #intrinsics: 1
 #tracks: 50840
 #residuals: 438129
 Initial RMSE: 0.305092
 Final RMSE: 0.188702
 Time (s): 17.3881

Usable motion priors: 1
Pose prior statistics (user units):
 - Starting median fitting error: 0.0649667
 - Final fitting error:
	 min: 0.0315291
	 mean: 0.096793
	 median: 0.0699325
	 max: 0.273009

-- Robust Resection of view: 39
  nfa=27.4825 inliers=7/8445 precisionNormalized=0.0269952 precision=229.827 (iter=0 ,sample=6880,1144,7649,7052,1075,8182,)
  nfa=26.3097 inliers=7/8445 precisionNormalized=0.00181329 precision=59.5649 (iter=2 ,sample=2351,1591,4619,8384,8086,8415,)
  nfa=25.1653 inliers=7/8445 precisionNormalized=0.000129922 precision=15.944 (iter=2 ,sample=2351,1591,4619,8384,8086,8415,)
  nfa=24.9485 inliers=7/8445 precisionNormalized=7.88136e-05 precision=12.4182 (iter=4 ,sample=8083,928,4100,6740,6759,2511,)
  nfa=-4.22006 inliers=677/8445 precisionNormalized=0.00883906 precision=131.51 (iter=5 ,sample=1198,41,3562,952,7733,5404,)
  nfa=-9.70317 inliers=291/8445 precisionNormalized=0.00308839 precision=77.7362 (iter=29 ,sample=3099,7209,6126,6925,4475,5617,)
  nfa=-10.8146 inliers=323/8445 precisionNormalized=0.00350706 precision=82.8378 (iter=51 ,sample=3226,3136,7166,6119,5184,6147,)
  nfa=-18.4524 inliers=393/8445 precisionNormalized=0.00429643 precision=91.6877 (iter=78 ,sample=7133,6179,4419,3189,3240,5741,)

-- Robust Resection 
-- Resection status: 1
-- #Points used for Resection: 8445
-- #Points validated by robust Resection: 393
-- Threshold: 91.6877

Bundle Adjustment statistics (approximated RMSE):
 #views: 1
 #poses: 1
 #intrinsics: 1
 #tracks: 393
 #residuals: 786
 Initial RMSE: 14.1153
 Final RMSE: 2.01159
 Time (s): 0.0808275


**- most recent develop branch of openMVG without -P switch**

-- Robust Resection of view: 26
  nfa=-3428.69 inliers=2489/3328 precisionNormalized=0.00610324 precision=111.443 (iter=0 ,sample=2711,451,3014,2779,426,3224,)
  nfa=-4341.19 inliers=2571/3328 precisionNormalized=0.00316729 precision=80.2815 (iter=0 ,sample=2711,451,3014,2779,426,3224,)
  nfa=-4799.67 inliers=3169/3328 precisionNormalized=0.00778623 precision=125.874 (iter=0 ,sample=2711,451,3014,2779,426,3224,)
  nfa=-5020.11 inliers=2974/3328 precisionNormalized=0.00436238 precision=94.2179 (iter=2 ,sample=205,3263,1956,1513,2688,2369,)
  nfa=-10483.8 inliers=2998/3328 precisionNormalized=6.84469e-05 precision=11.8018 (iter=5 ,sample=109,1627,2779,1186,896,440,)
  nfa=-11316.2 inliers=3194/3328 precisionNormalized=7.40726e-05 precision=12.2772 (iter=25 ,sample=1307,1741,250,1410,214,599,)
  nfa=-14346.4 inliers=3266/3328 precisionNormalized=1.12249e-05 precision=4.77929 (iter=26 ,sample=797,2759,2070,2934,746,141,)
  nfa=-14408 inliers=3182/3328 precisionNormalized=7.43231e-06 precision=3.88896 (iter=100 ,sample=455,1933,948,2872,1497,1013,)
  nfa=-14815.4 inliers=3232/3328 precisionNormalized=6.88695e-06 precision=3.74356 (iter=310 ,sample=810,901,1187,545,2174,156,)

-- Robust Resection 
-- Resection status: 1
-- #Points used for Resection: 3328
-- #Points validated by robust Resection: 3232
-- Threshold: 3.74356

Bundle Adjustment statistics (approximated RMSE):
 #views: 1
 #poses: 1
 #intrinsics: 1
 #tracks: 3232
 #residuals: 6464
 Initial RMSE: 1.10789
 Final RMSE: 0.441725
 Time (s): 0.072649

Bundle Adjustment statistics (approximated RMSE):
 #views: 174
 #poses: 19
 #intrinsics: 1
 #tracks: 50837
 #residuals: 438106
 Initial RMSE: 0.294899
 Final RMSE: 0.18902
 Time (s): 8.42316

-- Robust Resection of view: 39
  nfa=-17906.8 inliers=8030/8445 precisionNormalized=0.00150919 precision=54.3661 (iter=0 ,sample=6880,1144,7649,7052,1075,8182,)
  nfa=-19944.7 inliers=8187/8445 precisionNormalized=0.00100185 precision=44.2953 (iter=3 ,sample=4104,4364,2311,314,7944,831,)
  nfa=-20780.3 inliers=8353/8445 precisionNormalized=0.000963991 precision=43.4503 (iter=7 ,sample=1696,5956,1619,3072,319,4405,)
  nfa=-36712.8 inliers=8260/8445 precisionNormalized=1.0005e-05 precision=4.42655 (iter=10 ,sample=1522,5253,2473,182,349,6124,)
  nfa=-37585.6 inliers=8247/8445 precisionNormalized=7.63927e-06 precision=3.86796 (iter=336 ,sample=2245,5991,315,1425,7316,4786,)
  nfa=-38145 inliers=8250/8445 precisionNormalized=6.55171e-06 precision=3.58207 (iter=344 ,sample=3678,7306,116,4109,3964,6681,)

-- Robust Resection 
-- Resection status: 1
-- #Points used for Resection: 8445
-- #Points validated by robust Resection: 8250
-- Threshold: 3.58207

Bundle Adjustment statistics (approximated RMSE):
 #views: 1
 #poses: 1
 #intrinsics: 1
 #tracks: 8250
 #residuals: 16500
 Initial RMSE: 0.597018
 Final RMSE: 0.454781
 Time (s): 0.0888142


**- most recent develop branch of openMVG (and [#1253](https://github.com/openMVG/openMVG/pull/1253)) with -P switch**

-- Robust Resection of view: 26
  nfa=-3486.61 inliers=2568/3328 precisionNormalized=0.00685786 precision=118.132 (iter=0 ,sample=2711,451,3014,)
  nfa=-4309.52 inliers=2570/3328 precisionNormalized=0.00329093 precision=81.8335 (iter=0 ,sample=2711,451,3014,)
  nfa=-4803.04 inliers=3170/3328 precisionNormalized=0.00785557 precision=126.433 (iter=0 ,sample=2711,451,3014,)
  nfa=-11362.1 inliers=3182/3328 precisionNormalized=6.95418e-05 precision=11.8958 (iter=3 ,sample=3064,2890,2386,)
  nfa=-13357.8 inliers=3269/3328 precisionNormalized=2.33023e-05 precision=6.88606 (iter=55 ,sample=429,2557,1788,)

-- Robust Resection 
-- Resection status: 1
-- #Points used for Resection: 3328
-- #Points validated by robust Resection: 3269
-- Threshold: 6.88606

Bundle Adjustment statistics (approximated RMSE):
 #views: 1
 #poses: 1
 #intrinsics: 1
 #tracks: 3269
 #residuals: 6538
 Initial RMSE: 1.84987
 Final RMSE: 0.485007
 Time (s): 0.101834

Bundle Adjustment statistics (approximated RMSE):
 #views: 174
 #poses: 19
 #intrinsics: 1
 #tracks: 50818
 #residuals: 438161
 Initial RMSE: 0.297688
 Final RMSE: 0.189756
 Time (s): 9.89485

Usable motion priors: 1
Pose prior statistics (user units):
 - Starting median fitting error: 0.0639887
 - Final fitting error:
	 min: 0.0324055
	 mean: 0.0969153
	 median: 0.0704172
	 max: 0.272835

-- Robust Resection of view: 39
  nfa=18.3565 inliers=4/8445 precisionNormalized=0.0252895 precision=222.478 (iter=0 ,sample=6880,1144,7649,)
  nfa=18.2577 inliers=4/8445 precisionNormalized=0.0201413 precision=198.546 (iter=1 ,sample=7051,1073,8182,)
  nfa=15.9615 inliers=4/8445 precisionNormalized=0.000101729 precision=14.1104 (iter=1 ,sample=7051,1073,8182,)
  nfa=15.5233 inliers=5/8445 precisionNormalized=5.27974e-05 precision=10.1654 (iter=4 ,sample=2351,1591,4619,)
  nfa=14.6412 inliers=6/8445 precisionNormalized=3.45199e-05 precision=8.21963 (iter=8 ,sample=8083,928,4100,)
  nfa=-15.7279 inliers=666/8445 precisionNormalized=0.00864358 precision=130.066 (iter=10 ,sample=1198,41,3562,)
  nfa=-17.78 inliers=395/8445 precisionNormalized=0.00462331 precision=95.1249 (iter=17 ,sample=6135,4434,3301,)
  nfa=-20.9348 inliers=416/8445 precisionNormalized=0.00483799 precision=97.3083 (iter=41 ,sample=5743,4517,6604,)
  nfa=-24.8284 inliers=443/8445 precisionNormalized=0.00511957 precision=100.1 (iter=86 ,sample=3794,6915,4590,)
  nfa=-28.5392 inliers=420/8445 precisionNormalized=0.00469389 precision=95.8483 (iter=284 ,sample=6633,6989,7538,)

-- Robust Resection 
-- Resection status: 1
-- #Points used for Resection: 8445
-- #Points validated by robust Resection: 420
-- Threshold: 95.8483

Bundle Adjustment statistics (approximated RMSE):
 #views: 1
 #poses: 1
 #intrinsics: 1
 #tracks: 420
 #residuals: 840
 Initial RMSE: 23.3385
 Final RMSE: 6.08512
 Time (s): 0.0820006


**- v1.2 develop branch of openMVG () with -P switch**

-- Robust Resection of view: 26
  nfa=-2992.07 inliers=2482/3328 precisionNormalized=0.00913319 precision=136.327 (iter=0 ,sample=2711,451,3014,)
  nfa=-10244.5 inliers=3213/3328 precisionNormalized=0.000173615 precision=18.796 (iter=0 ,sample=2711,451,3014,)
  nfa=-15334.4 inliers=3220/3328 precisionNormalized=4.5365e-06 precision=3.03831 (iter=1 ,sample=2196,1666,1474,)
  nfa=-15466.6 inliers=3220/3328 precisionNormalized=4.11608e-06 precision=2.8941 (iter=18 ,sample=855,167,190,)
  nfa=-15740.9 inliers=3158/3328 precisionNormalized=2.49512e-06 precision=2.25329 (iter=27 ,sample=368,2406,1313,)
  nfa=-15847.3 inliers=3159/3328 precisionNormalized=2.31101e-06 precision=2.16856 (iter=61 ,sample=2838,2390,1170,)
  nfa=-15919.2 inliers=3144/3328 precisionNormalized=2.03034e-06 precision=2.03262 (iter=65 ,sample=2406,269,2747,)
  nfa=-15951.1 inliers=3171/3328 precisionNormalized=2.26387e-06 precision=2.14634 (iter=98 ,sample=1310,1998,1205,)
  nfa=-15974.3 inliers=3167/3328 precisionNormalized=2.18055e-06 precision=2.10647 (iter=182 ,sample=2855,1081,2979,)
  nfa=-15997 inliers=3136/3328 precisionNormalized=1.836e-06 precision=1.93289 (iter=388 ,sample=1267,283,921,)

-- Robust Resection 
-- Resection status: 1
-- #Points used for Resection: 3328
-- #Points validated by robust Resection: 3136
-- Threshold: 1.93289

Bundle Adjustment statistics (approximated RMSE):
 #views: 1
 #poses: 1
 #intrinsics: 1
 #tracks: 3136
 #residuals: 6272
 Initial RMSE: 0.397658
 Final RMSE: 0.374847
 Time (s): 0.165379

Bundle Adjustment statistics (approximated RMSE):
 #views: 174
 #poses: 19
 #intrinsics: 1
 #tracks: 50847
 #residuals: 438123
 Initial RMSE: 0.294252
 Final RMSE: 0.188743
 Time (s): 10.6828

Usable motion priors: 1
Pose prior statistics (user units):
 - Starting median fitting error: 0.0640137
 - Final fitting error:
	 min: 0.0326317
	 mean: 0.0966948
	 median: 0.0694979
	 max: 0.273369

-- Robust Resection of view: 39
  nfa=-39309.3 inliers=8155/8445 precisionNormalized=3.97186e-06 precision=2.78864 (iter=0 ,sample=6880,1144,7649,)
  nfa=-39493.7 inliers=8183/8445 precisionNormalized=3.96468e-06 precision=2.78612 (iter=7 ,sample=1949,3030,5451,)
  nfa=-39578 inliers=8097/8445 precisionNormalized=3.2968e-06 precision=2.54063 (iter=9 ,sample=5482,861,4876,)
  nfa=-39623.6 inliers=8228/8445 precisionNormalized=4.15081e-06 precision=2.85077 (iter=27 ,sample=3926,2872,6686,)
  nfa=-39937.1 inliers=8135/8445 precisionNormalized=3.1851e-06 precision=2.49722 (iter=49 ,sample=8303,4467,8342,)
  nfa=-40197.4 inliers=8054/8445 precisionNormalized=2.52861e-06 precision=2.22503 (iter=60 ,sample=3560,4341,6475,)
  nfa=-40395.5 inliers=8116/8445 precisionNormalized=2.68379e-06 precision=2.29229 (iter=64 ,sample=573,3668,8071,)
  nfa=-40589.2 inliers=8062/8445 precisionNormalized=2.28306e-06 precision=2.11424 (iter=96 ,sample=2222,4537,3548,)

-- Robust Resection 
-- Resection status: 1
-- #Points used for Resection: 8445
-- #Points validated by robust Resection: 8062
-- Threshold: 2.11424

Bundle Adjustment statistics (approximated RMSE):
 #views: 1
 #poses: 1
 #intrinsics: 1
 #tracks: 8062
 #residuals: 16124
 Initial RMSE: 0.424442
 Final RMSE: 0.420404
 Time (s): 0.117961

**Comment by rjanvier** [comment]
Hello @hcjghr,
I suggest you could use git bisect and log result to see when the regression happen.

**Comment by hcjghr** [comment]
Hi @rjanvier,

thanks for the suggestion. I've run it and determined that:

ca2a20718453cee3bd60f8daeaa07697123289f0 is the first bad commit
commit ca2a20718453cee3bd60f8daeaa07697123289f0
Author: pmoulon <pmoulon@gmail.com>
Date:   Sat Oct 7 10:05:01 2017 -0700

    Update to use bearing vector instead of point position #1133
    
    [cameras]
    - Fix Spherical model + Use Eigen in the ceres camera functor
    - Update some functions to deal with matrices instead of vector (ability to handle many point at once)
    
    [multiview] Update multiview relative pose computation module to use bearing vectors
    - multiview::MotionFromEssential exports now Poses instead of [R|t] couples.
    - update fundamental::kernel::EncodeEpipolarEquation to use bearing vectors.
    - EightPointRelativePoseSolver::Solve uses fundamental::kernel::EncodeEpipolarEquation.
    - Update the FivePoint relative pose estimation to use bearing vectors.
    - Update some multiview function to use Mat2X or Mat3X instead of Mat in order to have a more explicit API.
    - estimate_Rt_fromE => RelativePoseFromEssential
    
    [sfm]
    - Update triangulation routines in order to use bearing vector and camera poses.
    
    [samples]
    - Update the samples robust_essential_spherical to use either the 8point or the 5point essential algorithm to compute the essential matrix.
    
    [misc.]
    - Essential matrix ACKernelAdaptorEssential uses FivePointSolver instead of FivePointKernel.

:040000 040000 43fe203246df35cd6bcda9abbb4fd79e7593f343 e46861a11f177933c6c5b3f706e98774b4ec4bb1 M      src

Now I guess its time to dig deep and find the problem

**Comment by hcjghr** [comment]
Hi,

I looked into the problem a bit but the commit is so big and complex that I wasn't able to find the specific problem. However I'm still noticing the same problem (with the newest code) so I wanted to ask if anybody faced the same problem. @pmoulon any idea?

Thanks

**Comment by hcjghr** [comment]
Hi,

As I came across the same problem again I wanted to ask if anyone (@pmoulon, @rjanvier) has experienced similar problem and how to deal with it?

Thanks

**Comment by rjanvier** [comment]
Hello, I haven't tested but I do agree with you, it's really problematic. Maybe you could try two or three thing. 
- be sure nothing changed in the geodesy module in the commit that introduce the regression
- Since it started with the use of bearing vector, you could try to use the old resection scheme in the current branch: In the localization module you should consider you always use pinhole camera's and you use the old AC-RANSAC resection kernel for pinhole camera instead of the new generalized one 
- If it still do not work as expected, I think we should investigate in the BA module

**Comment by rjanvier** [comment]
Ok, It was what I suspected. Apart your bug, I also noticed strange behavior since 1.3: the AC-RANSAC threshold tends to "drift" across the reconstruction. You can end with very high threshold like 150 px even if the visual results look as good as ones from 1.2. I agree it could be a normalization problem, but last time I checked (for another purpose) it looks ok. On this basis, do you think you can continue to figure out what's really going on?

**Comment by hcjghr** [comment]
Hi,

I can confirm that I have also been noticing the drift in the threshold in multiple occasions (with and without the gps). To be fair I didnt paid much attention as the results were still on majority of cases ok, but sometimes I got better results with the 1.2 version. 

I tried to investigate a bit more today but got quite lost in all the details, especially as my initial guess that it would simply work with the old "pinhole" assumption for acransac kernel didnt work out. I get strange results later with a lot of points removed by outlier rejection schemes. I'm away for the next two weeks so I cant do much but after that I can try to see if I can figure something out...but i think it will be quite hard without input of @pmoulon...until then I guess its back to 1.2 version :( 

I think I've shared a dataset with both you and pmoulon that produced the error, but if you cant find it I can send it again if it will help.

I will keep you posted and thanks for the help so far!
